### PR TITLE
Warehouse

### PR DESCRIPTION
**Pull request by dronaka**

### CODE REVIEW

File: warehouse-service/api.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+from dbworkers.SQLAlchemySession import SQlAlchemySession
+from falcon.status_codes import HTTP_400, HTTP_200
+
+
+@hug.get("/api/warehouse/items")
+def GetItems(session: SQlAlchemySession):
+
+    result = []
+    for instanse in session.query(FreeItemsQuantity):
+        item = session.query(ItemType).get(instanse.itemTypeId)
```
Comment (BlooDCrescent) [inline warehouse-service/api.py]: Ask Anton if we can get ALL items in one batch. If we make a request for each item instance, it'll create an immense load on our db

File: warehouse-service/api.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    return ItemDto(id=item.id, name=item.name, price=item.price, amount=quantity)
+
+
+@hug.put("/api/warehouse/items/{itemTypeId}/additional/{amount}")
+def AddExistingItems(session: SQlAlchemySession, itemTypeId: int, amount: int):
+    item = session.query(FreeItemsQuantity).filter_by(itemTypeId=itemTypeId).first()
+    itemType = session.query(ItemType).get(item.itemTypeId)
+    print(item.quantity)
+    print(amount)
+    item.quantity = item.quantity + amount
```
Comment (BlooDCrescent) [inline warehouse-service/api.py]: Is this approach correct? Shouldn't you load an item back to database before flushing session?

File: warehouse-service/api.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        id=itemType.id, name=itemType.name, price=itemType.price, amount=item.quantity
+    )
+
+
+@hug.post("/api/warehouse/items")
+def CreateItems(session: SQlAlchemySession, body):
+    itemCreationDto = ItemCreationDto(
+        name=body["name"], amount=body["amount"], price=body["price"]
+    )
+    new_itemType = ItemType(itemCreationDto.name, itemCreationDto.price)
```
Comment (BlooDCrescent) [inline warehouse-service/api.py]: What should we do if there's already an item with the same name in the database?

File: warehouse-service/dbworkers/path.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import hug
```
Comment (BlooDCrescent) [inline warehouse-service/dbworkers/path.py]: Dafuq?

Comment (dronaka) [reply warehouse-service/dbworkers/path.py]: блэт... забыл убрать этот файл

File: warehouse-service/api.py (hunk 2 above)
Comment (dronaka) [reply warehouse-service/api.py]: ну не знаю, всё работает

File: warehouse-service/client_api/api.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    response = requests.get(request_url)
+
+    return ItemDto(**response.json())
+
+
+def AddExistingItems(itemTypeId: int, amount: int):
+    request_url = "/api/warehouse/items/{itemTypeId}/additional/{amount}" % (
+        itemTypeId,
+        amount,
+    )
```
Comment (toshiks) [inline warehouse-service/client_api/api.py]: It should be with hostname. And you should fix bug with format string.

File: warehouse-service/models/FreeItemsQuantity.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from models.base import Base
+from sqlalchemy import Column, Integer
+
+
+class FreeItemsQuantity(Base):
+    __tablename__ = "FreeItemsQuantity"
+
+    id = Column(Integer, primary_key=True)
+    itemTypeId = Column(Integer)
```
Comment (toshiks) [inline warehouse-service/models/FreeItemsQuantity.py]: In python you should write: item_type_id. Fix that in all files.

File: warehouse-service/api.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+from falcon.status_codes import HTTP_400, HTTP_200
+
+
+@hug.get("/api/warehouse/items")
+def GetItems(session: SQlAlchemySession):
+
+    result = []
+    for free_item, item_type in (
+        session.query(FreeItemsQuantity, ItemType).join(ItemType).all()
+    ):
```
Comment (toshiks) [inline warehouse-service/api.py]: I think more readable will be if you save  session.query(FreeItemsQuantity, ItemType).join(ItemType).all()
And I don't understand what does it do?

File: warehouse-service/api.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+@hug.get("/api/warehouse/items/{item_type_id}")
+def GetItemsByID(session: SQlAlchemySession, item_type_id: int):
+
+    item = session.query(ItemType).get(item_type_id)
+    quantity = (
+        session.query(FreeItemsQuantity)
+        .filter_by(item_type_id=item.id)
+        .first()
+        .quantity
+    )
```
Comment (toshiks) [inline warehouse-service/api.py]: It isn't good way to use SQLAlchemy. Try to read the article: https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html 
And use **relationship**

File: warehouse-service/api.py (hunk 7 above)
Comment (dronaka) [reply warehouse-service/api.py]: я тоже не понимаю. Он должен возвращать объединенную  таблицу, но не возвращает.

File: warehouse-service/models/ReservedItemsQuantity.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from models.base import Base
```
Comment (BlooDCrescent) [inline warehouse-service/models/ReservedItemsQuantity.py]: I thought we discussed that reserved items should be under the Order service

Comment (dronaka) [reply warehouse-service/models/ReservedItemsQuantity.py]: да ты прав. Просто файлик лежит, ждёт своего часа.

**Pull request merged**